### ISSUE TITLE
HLSL/WebGPU Binding Model Mismatch

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#192]
@litherum discovered in #188 that there is a mismatch between the current WebGPU binding model (based on Vulkan) and the to-be-blessed WHLSL - assuming it matches D3D's HLSL.

As I understand it, there are two options:
* Bindings are keyed on a single int, and the language requires that `space0`, etc., be specified in the input declaration. In #188, @magcius explained that DXC does this for Vulkan today.
* Bindings are keyed on the tuple `(binding type, int)`, and shaders use t0, s0, u0, b0 without space0. In this case we may want to use the strings "t0" etc as the keys (rather than two actual dictionary fields).

We need to discuss this soon to unblock any changes to the bind group IDL.

This may be more nuanced than I have captured here. We probably talked about some of this during the binding model investigations (where we were talking about operating _on top_ of d3d). There is useful info there too: #19

**Comment by magcius** [comment]
Not quite. There's three spaces:

1. Bind group ("descriptor set" in Vulkan terminology). This doesn't exist in D3D11/D3D12. The former has a global set of per-shader-stage bindings that can be set dynamically per-draw-call, and the latter has a fully bindless model with root descriptor sets. This is declared through `space0` / `space1` in Spiregg.

2. Register type. HLSL, for legacy D3D11 reasons, uses *separate* groups for resource types. That is, in HLSL, register b0 ("Buffer 0") is a different slot than s0 ("Sampler 0"). Spiregg emulates by making the user pass `-fvk-b-shift 256 -fvk-s-shift 512`, which will map b0 to binding index 256, and s0 to binding index 512. There are five such types: b (Buffer), t (Texture), s (Sampler), c (Buffer offset, which I believe doesn't exist in SPIR-V), and u (Unordered Access View, which is similar to an SSBO or image load/store in GL/Vulkan).

3. Register number. This is the `0` in `b0`. Spiregg adds this to the specified shift for the register type.

There's also a fourth, bonus component here: Per-stage bindings, something I believe D3D11 supports but Vulkan does not. Part of the `register` syntax is that you can do `register(ps, b0)` to specify that this binding only exists in the pixel shader, or `register(vs, b0)` for vertex shader. Spiregg doesn't emulate this -- you can do it yourself through manipulation of the -fvk-b-shift, etc. arguments as you compile the different shaders.

**Comment by litherum** [comment]
The most natural way is for `register(t3, space4)` means “the texture with binding=3 in the 4th bindgroup”

**Comment by Kangz** [comment]
@litherum's suggestion sounds good. That's also how SPIRV group and offset decorations are translated by SPIRV-Cross. I'll assume this is resolved and close but feel free to reopen!

**Comment by kvark** [comment]
@Kangz that doesn't answer the question of whether it would be allowed for WHLSL to have multiple different bindings at the same index, i.e.:
```
register(t3, space2) ...
register(s3, space2) ...
```

**Comment by devshgraphicsprogramming** [comment]
We may need to consider for a moment the implications of taking each route.
https://github.com/gpuweb/gpuweb/issues/19

if we allow for the bind model to be D3D12 like, a table of 3-tuple of (resource_type,index,set) or possibly even a 4-tuple of (resource_type,index,set,stage) has to be remapped to 4 tables of (resource_type,set) . Whenever the input set of tuples is not identical then the remapping will not be indentical as well, resulting in whenever D3D12 would have to change root signature Vulkan will have to probably bind all descriptor sets anew as well (or end up using only 1 descriptor set).

however if you remap (index,set) to D3D12 you will have to waste some slots, **but not actual D3D12 Root Signature space** because D3D12 is bindless. Whenever Vulkan changes one descriptor set you have to update the root table or change a signature, but because D3D12 targets desktop GPUs (most of which have descriptor indexing in Vulkan, and some even full bindless) this should have nowhere near as much of a cost as with doing unneeded Vulkan desc set rebinds. Plus you'd have to change these those signatures/tables anyway if anything about your descriptor sets were to change! 

And also there's that thing where you dont need to rebind a descriptor set if the layout matches.

I do this sort of remapping with OpenGL from Vulkan and it hurts the number of slots I can use but I trade slot space for being able to not rebind the objects in OpenGL slots corresponding to descriptor sets that were not rebound. D3D12 unlike OpenGL does not suffer a "wasted binding index problem".

To quote @Kangz 
> * Translating from D3D12 to Vulkan would require runtime creation of descriptor sets when a root descriptor is set.
> * Vulkan requires the number of each type of descriptor for descriptor pool creation, whereas D3D12 requires the total number of descriptor for descriptor heap creation.

With Indirect Argument Buffers, Metal should be able to do as well as Vulkan, although I'm no expert on Apple APIs

TL;DR You're likely to have far more "slots" and "active slots" for shader resources on the most constrained D3D12 device (probably some pre 600 series Intel HD) than "slots" in a descriptor set (layout) on the most constrained Vulkan Device (mobile).

(index,space/set) is a better choice.

**Comment by devshgraphicsprogramming** [comment]
Also from my personal experience as an ISV I find the concept (index,set) more valuable than (index,type) {although I could live with (index,set,type)}.

This is because it makes my renderer architecture much more simple, I assign:
set 0: MeshBuffer/Asset specific data such as material textures, etc.
set 1: Instance/Node specific data such as world/normal transforms, material data that is specific to each instance, etc.
set 2: RenderView (camera) specific data such as view transforms, renderpass constants, etc.
set 3: Global/Per-Frame constants such as tonemapper/autoexposure parameters, seeds, timers etc.

If I didn't have a "set"-based binding system I'd probably implement my own, but that would involve painful caching of WebGPU Bind Group (assuming you backpedal and expose only 1 instead of 4) per-asset-per-node-per-camera-per-frame

**Comment by kainino0x** [comment]
More discussion is happening on 188, so you might want to post a link to your comment there, or chime in on that discussion separately.

**Comment by devshgraphicsprogramming** [comment]
Doesn't it render this issue a duplicate?

**Comment by kvark** [comment]
@devshgraphicsprogramming thank you for greatly details ISV input!

I have to note that it's not D3D12 root signatures versus Vulkan descriptor sets models that are on the table. Nobody in the group suggested to use the root signatures "as-is". Instead, on the binding model we are going to have the resource groups either way. The question is rather - how they map to the shader:
  1. the bind group layout describes mapping into shader registers aka D3D11 HLSL: s3, t1, etc
  2. the bind group layout describes mapping from descriptor indices to descriptors *within the set/group*, and the shader has to explicitly specify the set/group.

With this in mind, the performance issues of D3D12 to Vulkan are not relevant. Either of the mapping would not require D3D12 or Vulkan to rebind any other descriptors than the one belonging to the resource group the user attempts to bind.

Also, use-case you described would work equally well with both of those bind models. It's really just a question of how the resource interface is described in the shaders at the moment.
